paddle/fluid/operators/math/softmax_impl.h

```diff
@@ -65,6 +65,39 @@ void SoftmaxFunctor<DeviceContext, T>::operator()(const DeviceContext& context,
                                                  .broadcast(one_by_class));
 }
 
+template <typename DeviceContext, typename T>
+class SoftmaxFunctor<DeviceContext, T, true> {
+  void operator()(const DeviceContext& context, const framework::Tensor* X,
+                  framework::Tensor* Y) {
+    auto logits = EigenMatrix<T>::From(*X);
+    auto softmax = EigenMatrix<T>::From(*Y);
+
+    const int kBatchDim = 0;
+    const int kClassDim = 1;
+
+    const int batch_size = logits.dimension(kBatchDim);
+    const int num_classes = logits.dimension(kClassDim);
+
+    Eigen::DSizes<int, 1> along_class(kClassDim);
+    Eigen::DSizes<int, 2> batch_by_one(batch_size, 1);
+    Eigen::DSizes<int, 2> one_by_class(1, num_classes);
+
+    auto shifted_logits = (logits -
+                           logits.maximum(along_class)
+                               .eval()
+                               .reshape(batch_by_one)
+                               .broadcast(one_by_class));
```

Comment:
The only difference between `class SoftmaxFunctor<DeviceContext, T, is_test>` and `class SoftmaxFunctor<DeviceContext, T, true>` is line 57: `unaryExpr(ValueClip<T>())`.

Thus, could you combine `class SoftmaxFunctor<DeviceContext, T, true>` into `class SoftmaxFunctor<DeviceContext, T, is_test`, to make the coder much cleaner and less?

Reply:
@luotao1 You are right, but I 'm working on further changes (MKL replacing Eigen + omp vectorization) that will change this code much more

Reply:
Got it

Reply:
To make it a bit clearer, I'm planning to add MKL+vectorization in Next PR. So after review is done I would like this one merged.

Reply:
Hope the Next PR! 
By the way, when will you enhance https://github.com/PaddlePaddle/Paddle/pull/14053#discussion_r229623740?

Reply:
@luotao1 Sorry for delay with Max sequence pool followup PR. Once I'm done with softmax I will return to #14053 to fix/test suggesstions.